File: pos-scanner-web/backend_fastapi/app/models/scan_result.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, ForeignKey
from sqlalchemy.orm import relationship

from app.db.session import Base


class ScanResult(Base):
    """Scan result model."""

    __tablename__ = "scan_results"

    id = Column(Integer, primary_key=True, index=True)
    ip = Column(String(50), nullable=False)
    merchant_id = Column(String(100), index=True)
    name = Column(String(200))
    version = Column(String(50))
    type = Column(String(50))
    full_data = Column(Text)
    scanned_at = Column(DateTime, default=datetime.utcnow)
    is_online = Column(Boolean, default=True)
    last_online_time = Column(DateTime, default=datetime.utcnow)
    owner_id = Column(Integer, ForeignKey("users.id"))

    def to_dict(self):
        return {
            "ip": self.ip,
            "merchantId": self.merchant_id,
            "name": self.name,
            "version": self.version,
            "type": self.type,
            "status": "success" if self.merchant_id else "error",
            "fullData": self.full_data,
            "isOnline": self.is_online,
            "lastOnlineTime": self.last_online_time.isoformat() if self.last_online_time else None,
            "ownerId": self.owner_id,
        }


class DeviceClaim(Base):
    """Device claim request model."""

    __tablename__ = "device_claims"

    id = Column(Integer, primary_key=True, index=True)
    merchant_id = Column(String(100), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    status = Column(String(20), default="pending")  # pending, approved, rejected
    created_at = Column(DateTime, default=datetime.utcnow)
    processed_at = Column(DateTime, nullable=True)
    processed_by = Column(Integer, ForeignKey("users.id"), nullable=True)

    def to_dict(self):
        return {
            "id": self.id,
            "merchantId": self.merchant_id,
            "userId": self.user_id,
            "status": self.status,
            "createdAt": self.created_at.isoformat() if self.created_at else None,
            "processedAt": self.processed_at.isoformat() if self.processed_at else None,
        }


class ScanSession(Base):
    """Scan session model - records last scan time."""

    __tablename__ = "scan_sessions"

    id = Column(Integer, primary_key=True, default=1)
    last_scan_at = Column(DateTime, default=datetime.utcnow)

File: pos-scanner-web/backend-go/internal/models/scan_result.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type ScanResult struct {
	ID             uint           `gorm:"primaryKey" json:"id"`
	IP             string         `gorm:"size:50;not null" json:"ip"`
	MerchantID     *string        `gorm:"size:100;index" json:"merchant_id"`
	Name           *string        `gorm:"size:200" json:"name"`
	Version        *string        `gorm:"size:50" json:"version"`
	Type           *string        `gorm:"size:50" json:"type"`
	FullData       *string        `gorm:"type:text" json:"full_data"`
	ScannedAt      time.Time      `json:"scanned_at"`
	IsOnline       bool           `gorm:"default:true" json:"is_online"`
	LastOnlineTime time.Time      `json:"last_online_time"`
	OwnerID        *uint          `gorm:"index" json:"owner_id"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

func (ScanResult) TableName() string {
	return "scan_results"
}

func (s *ScanResult) ToDict() map[string]interface{} {
	merchantID := ""
	if s.MerchantID != nil {
		merchantID = *s.MerchantID
	}

	status := "error"
	if merchantID != "" {
		status = "success"
	}

	result := map[string]interface{}{
		"ip":              s.IP,
		"merchantId":      merchantID,
		"name":            s.Name,
		"version":         s.Version,
		"type":            s.Type,
		"status":          status,
		"fullData":        s.FullData,
		"isOnline":        s.IsOnline,
		"lastOnlineTime":  s.LastOnlineTime.Format(time.RFC3339),
		"ownerId":         s.OwnerID,
	}
	return result
}
